File: day5/puzzle2.py
from sys import exit


def dest_to_source(dest: int, map: list[tuple[int, int, int]]) -> int:
    for r in map:
        if dest >= r[0] and dest < r[0] + r[2]:
            return r[1] + (dest - r[0])
    return dest


def main():
    """Application entry-point."""
    with open("puzzle_input.txt", "rt") as f:
        seed_ranges: list[tuple[int, int]] = []
        maps: list[list[tuple[int, int, int]]] = []
        map_index = -1
        for line in f:
            line = line.strip()
            if len(line) == 0:
                continue
            if line.startswith("seeds: "):
                values = [int(seed) for seed in line[7:].split(" ")]
                for i in range(int(len(values) / 2)):
                    seed_ranges.append((values[2 * i], values[2 * i + 1]))
            else:
                if line.endswith(" map:"):
                    map_index += 1
                    maps.append([])
                else:
                    values = [int(v) for v in line.split(" ")]
                    maps[map_index].append((values[0], values[1], values[2]))
    location = -1
    while True:
        location += 1
        seed = None
        dest = location
        for map_index in range(6, -1, -1):
            source = dest_to_source(dest, maps[map_index])
            if not source:
                break
            dest = source
        if not source:
            continue
        for seed_range in seed_ranges:
            if source >= seed_range[0] and source < seed_range[0] + seed_range[1]:
                seed = source
                print(f"done - seed = {seed}, location = {location}")
                exit(0)


if __name__ == "__main__":
    main()

File: day5/puzzle2.lua
local function dest_to_source(dest, map)
    for _, r in ipairs(map) do
        if dest >= r[1] and dest < r[1] + r[3] then
            return r[2] + (dest - r[1])
        end
    end
    return dest
end

local function main()
    local f = io.open("puzzle_input.txt", "r")
    if not f then
        os.exit(1)
    end
    local line_no = 0
    local seed_ranges = {}
    local maps = {}
    local map_index = 0
    for line in f:lines() do
        line_no = line_no + 1
        if #line > 0 then
            if line:sub(1, 7) == "seeds: " then
                local values = {}
                for value in line:sub(8, -1):gmatch("%d+") do
                    table.insert(values, tonumber(value))
                end
                for i = 0, (#values / 2) - 1 do
                    table.insert(seed_ranges, { values[2 * i + 1], values[2 * i + 2] })
                end
            elseif line:sub(-5, -1) == " map:" then
                map_index = map_index + 1
                table.insert(maps, {})
            else
                local values = {}
                for value in line:gmatch("%d+") do
                    table.insert(values, tonumber(value))
                end
                table.insert(maps[map_index], values)
            end
        end
    end
    f:close()
    local location = -1
    local found = false
    while not found do
        location = location + 1
        local dest = location
        for map_index = 7, 1, -1 do
            dest = dest_to_source(dest, maps[map_index])
        end
        for _, seed_range in ipairs(seed_ranges) do
            if dest >= seed_range[1] and dest < seed_range[1] + seed_range[2] then
                print(string.format("seed = %d, location = %d", dest, location))
                found = true
                break
            end
        end
    end
end

main()
